MicroBenchmarks/LoopVectorization/LoopInterleaving.cpp

```diff
@@ -0,0 +1,399 @@
+// This program tests performance impact of Interleaving Count with varying loop
+// iteration count for different types of loops, such as loops with or
+// without reductions inside it, loops with different vectorization widths.
+#include <iostream>
+#include <memory>
+#include <random>
+
+#include "benchmark/benchmark.h"
+
+#define ELEMENTS 2048
+#define ALIGNED16 __attribute__((aligned(16)))
+
+static std::mt19937 rng;
+unsigned int g_sum = 0;
+
+int A[ELEMENTS] ALIGNED16;
+int B[ELEMENTS] ALIGNED16;
+int C[ELEMENTS] ALIGNED16;
+int D[ELEMENTS] ALIGNED16;
+int E[ELEMENTS] ALIGNED16;
+int F[ELEMENTS] ALIGNED16;
+
+// Initialize arrays with random numbers.
+static void init_data(unsigned N) {
+  std::uniform_int_distribution<int> distrib(std::numeric_limits<int>::min(),
+                                             std::numeric_limits<int>::max());
+  for (unsigned I = 0; I < N; I++) {
+    A[I] = distrib(rng);
+    B[I] = distrib(rng);
+    C[I] = distrib(rng);
+    D[I] = distrib(rng);
+    E[I] = distrib(rng);
+    F[I] = distrib(rng);
+  }
+}
+
+static void __attribute__((always_inline))
+runBenchForLoopInterleaving(benchmark::State &state, int (*Fn)(int),
+                            int Iterations) {
+  std::uniform_int_distribution<int> distrib(std::numeric_limits<int>::min(),
+                                             std::numeric_limits<int>::max());
+  init_data(ELEMENTS);
+  for (auto _ : state) {
+    benchmark::DoNotOptimize(A);
+    benchmark::DoNotOptimize(B);
+    benchmark::DoNotOptimize(C);
+    benchmark::DoNotOptimize(D);
+    benchmark::DoNotOptimize(E);
+    benchmark::DoNotOptimize(F);
+    benchmark::ClobberMemory();
+    g_sum += Fn(Iterations);
+  }
+}
+
+#define STRINGIFY(a) #a
+
+// Loops without Reduction with different vectorization configurations
+
+static int __attribute__((noinline)) loopNoReductionAutoVec(int Iterations) {
+  for (int J = 0; J < Iterations; J++) {
```

Comment:
Could you double check that this loop doesn't get unrolled in case LLVM can propagate the constant iterations counts through to the function (may happen even with `no inline`)

Reply:
I checked that it doesn't unroll the loop because of propagating constant iteration counts & specifically disabled loop unrolling for the auto-vectorization cases.